src/subsystems/nodejs/builders/strict-builder/nodejs_builder/nodejs_builder/lib/dependencies.py

```diff
@@ -0,0 +1,135 @@
+from dataclasses import dataclass
+from typing import Any, Callable, Literal, Optional, TypedDict, Union
+
+from .logger import logger
+
+
+@dataclass
+class Dependency:
+    name: str
+    version: str
+    derivation: str
+    parent: Union["Dependency", None] = None
+    dependencies: Union[dict[str, Any], None] = None
+
+    def repr(self: "Dependency") -> str:
+        return f"{self.name}@{self.version}"
+
+
+def get_all_deps(all_deps: dict[str, Any], name: str, version: str) -> list[str]:
+    """
+    Returns all dependencies. as flattened list
+    """
+
+    def is_found(acc: Any, dep: Dependency, dep_tree: Optional[DepsTree]) -> bool:
+        return not bool(acc)
+
+    def find_exact_dependency(
+        acc: Any, dep: Dependency, dep_tree: Optional[DepsTree]
+    ) -> Any:
+        if acc:
+            return acc
+
+        if dep.repr() == f"{name}@{version}":
+            return dep_tree
+        return None
+
+    subtree = recurse_deps_tree(
+        all_deps, find_exact_dependency, acc=None, pred=is_found, order="top-down"
+    )
+
+    def flatten(acc: Any, dep: Dependency, dep_tree: Optional[DepsTree]) -> Any:
+        acc.append(dep.repr())
+        return acc
+
+    flattened: list[str] = []
+    if subtree:
+        flattened = recurse_deps_tree(subtree, flatten, [])
+
+    return flattened
+
+
+class Meta(TypedDict):
+    derivation: str
+    deps: Optional[dict[str, dict[str, Any]]]
+
+
+DepsTree = dict[str, dict[str, Meta]]
+
+
+def recurse_deps_tree(
+    deps: DepsTree,
+    cb: Callable[[Any, Dependency, Optional[DepsTree]], Any],
+    acc: Any,
+    parent: Union[Dependency, None] = None,
+    order: Literal["bottom-up", "top-down"] = "bottom-up",
+    pred: Optional[Callable[[Any, Dependency, Optional[DepsTree]], bool]] = None,
+):
+    """
+    Generic function that traverses the dependency tree and calls
+    'cb' on every node in the tree
+
+    Parameters
+    ----------
+    deps : DepsTree
+        The nested tree of dependencies, that will be iterated through.
+    cb : Callable[[Any, Dependency, Optional[DepsTree]], Any]
+        takes an accumulator (like 'fold' )
+    acc : Any
```

Comment:
Meaning of very short variable names is hard to grasp when reading the code.

Reply:
renamed it to `callback`